### DISCUSSION TITLE
HMR with parcel watch?

### DISCUSSION THREAD

**Discussion by mikedavies-dev** [parcel-bundler/parcel#8099] [Q&A]
Hi all, I have a bit of a strange setup and I'm struggling to get HMR working.

I'm using Parcel to compile TS/React files for use in a WordPress plugin and I use MAMP PRO (https://www.mamp.info/en/mamp-pro/mac/) to host my development environment.

The JS files are compiled by parcel into the plugin's JS folder and loaded from the browser but the WebSocket connection for HML always fails.

I've tried two setups:

### 1. Using `parcel watch`
When using `parcel watch` to compile the code, the browser tries to connect to the localhost on port `1234` but it seems that `parcel watch` doesn't start the development server so the connection fails.

I'm sure I'm missing something because the docs mention that HMR should work with `parcel watch` but I can't see how. I'm not using html files but compiling JS into modules using the `targets` definition in my `package.json` file so maybe that is stopping the HMR server from starting? Or is there some other command to start the server?

Are there any docs for using HMR with `parcel watch`?

### 2. Using `parcel serve`
Using `parcel serve` starts the development server but because I'm connecting to WordPress on a different server port the browser fails to connect to WebSocket even if I specifically provide the HMR port using `--hml-port`.

The issue seems to be the `HMR_PORT` definition in `@parcel/runtime-browser-hmr/lib/HMRRuntime.js`, specifically:

```javascript
`var HMR_PORT = ${JSON.stringify(port != null && ( // Default to the HTTP port in the browser, only override
      // in watch mode or if hmr port != serve port
      !options.serveOptions || options.serveOptions.port !== port) ? port : null)};
```
If the HMR port is the same as the serve port then `HMR_PORT` is set to `null` so the browser uses the current port (i.e. the WordPress port) so the connection fails. But if I change the HMR port is set to something different to the serve port then `HMR_PORT` is used but the connection fails because the development server is running on a different port!

If I force the `HMR_PORT` value myself it works fine, i.e.

```javascript
var HMR_PORT = ${JSON.stringify(port)};
```

I suppose I could setup a proxy server running on the HMR port that redirects to the development port but that seems like overkill :)

Can anyone suggest a better way of setting this up?

**Comment by mikedavies-dev** [comment]
I've been digging into this a little more and it looks like Parcel does start the HMR server if `--hmr-port` is provided but it isn't started with HTTS/SSL and because my page is served over HTTPS, JS bundle tries to connect via `wss` and fails.

So I guess I need either a way to force `HMR_PORT` to be the port instead of `null` or I need a way to start the HMR server on HTTPS if passed the `--https` flag.

The first option seems the easiest with an extra flag like `--hmr-force-port` or something but I'm not sure that would be useful others. The code would look something like this after adding the new CLI option:

```javascript
const {host, port, forcePort} = options.hmrOptions;

const hrmPort = 
  port != null &&
    // Default to the HTTP port in the browser, only override
    // in watch mode or if hmr port != serve port
    (!options.serveOptions || options.serveOptions.port !== port || forcePort)
    ? port
    : null;
```

I'd be happy to create a PR for the above `--hmr-force-port` or similar if it's something that would be useful/accepted?

**Comment by JeremyBYU** [comment]
@ant-fx  I was running into nearly this exact same issue. In my case I was using Flask as the web server (on port 5000) and had it proxy to the parcel dev webserver (port 1234)  for all the static assets (html, css, javascript). See details [here](https://ajhyndman.medium.com/hot-reloading-with-react-and-flask-b5dae60d9898). The article has almost my same setup except I am using parcel instead of webpack.  However, HMR would not work because it was a websocket connection *and* the port was defaulting to 5000, not the 1234 of the parcel webserver.  I believe the HMR port was defaulting to null because it was the *same* as the serving port, and as you pointed out from the code snippet, it will default to null in that case.

I added the flag `--hmr-port 1235` to force HMR to be served on a different port. It now works for me. My understanding is that static assets will be served from port 1234 now and hmr (websockets) will now be served on port 1235. I know you mentioned that this did **not** fix it for you:

> But if I change the HMR port is set to something different to the serve port then HMR_PORT is used but the connection fails because the development server is running on a different port!

I am not sure why that doesn't work for you. I thought I would share my sucesss and the blog post and see if any of this helps.
